ogr/ogrsf_frmts/jsonfg/ogrjsonfgreader.cpp

```diff
@@ -0,0 +1,1260 @@
+            oParser.ExceptionOccurred())
+        {
+            return false;
+        }
+        if (oParser.IsTypeKnown() && !oParser.IsFeatureCollection())
+        {
+            break;
+        }
+        if (bFinished)
+            break;
+    }
+
+    if (!oParser.IsTypeKnown() || !oParser.IsFeatureCollection())
+    {
+        fp->Seek(0, SEEK_END);
+        const vsi_l_offset nFileSize = fp->Tell();
+        const vsi_l_offset nRAM =
+            static_cast<vsi_l_offset>(CPLGetUsablePhysicalRAM());
+        if (nRAM == 0 || nRAM > nFileSize * 20)
+        {
+            // Only try full ingestion if we have 20x more RAM than the file
+            // size
+            bCanTryWithNonStreamingParserOut = true;
+        }
+        return false;
+    }
+
+    poObject_ = oParser.StealRootObject();
+
+    return FinalizeGenerateLayerDefns(true);
+}
+
+/************************************************************************/
+/*                OGRJSONFGReader::GenerateLayerDefns()                 */
+/************************************************************************/
+
+bool OGRJSONFGReader::GenerateLayerDefns()
+{
+    const GeoJSONObject::Type objType = OGRGeoJSONGetType(poObject_);
+    if (objType == GeoJSONObject::eFeature)
+    {
+        if (!GenerateLayerDefnFromFeature(poObject_))
+            return false;
+    }
+    else if (objType == GeoJSONObject::eFeatureCollection)
+    {
+        json_object *poObjFeatures =
+            OGRGeoJSONFindMemberByName(poObject_, "features");
+        if (nullptr != poObjFeatures &&
+            json_type_array == json_object_get_type(poObjFeatures))
+        {
+            const auto nFeatures = json_object_array_length(poObjFeatures);
+            for (auto i = decltype(nFeatures){0}; i < nFeatures; ++i)
+            {
+                json_object *poObjFeature =
+                    json_object_array_get_idx(poObjFeatures, i);
+                if (!GenerateLayerDefnFromFeature(poObjFeature))
+                {
+                    return false;
+                }
+            }
+        }
+        else
+        {
+            CPLError(CE_Failure, CPLE_AppDefined,
+                     "Invalid FeatureCollection object. "
+                     "Missing \'features\' member.");
+            return false;
+        }
+    }
+    else
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "Missing or unhandled root type object");
+        return false;
+    }
+
+    return FinalizeGenerateLayerDefns(false);
+}
+
+/************************************************************************/
+/*             OGRJSONFGReader::FinalizeGenerateLayerDefns()            */
+/************************************************************************/
+
+bool OGRJSONFGReader::FinalizeGenerateLayerDefns(bool bStreamedLayer)
+{
+    json_object *poName = CPL_json_object_object_get(poObject_, "featureType");
+    // The spec allows an array of strings, but we don't support that
```

Comment:
Should we raise a warning in this case? To inform users of the reason for the failure.

Reply:
done